In [1]:
import pandas as pd
data = pd.read_csv("./train.csv")
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [2]:
test_data = pd.read_csv("./test.csv")
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

In [3]:
data.shape

(1460, 81)

In [4]:
data.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [5]:
test_data.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      227
LotArea            0
                ... 
MiscVal            0
MoSold             0
YrSold             0
SaleType           1
SaleCondition      0
Length: 80, dtype: int64

In the PoolQC I have 1453 missing points from 1460 so i decided to remove this feature

In [6]:
data = data.drop('PoolQC', axis=1)
test_data = test_data.drop('PoolQC', axis=1)
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... 3SsnPorch ScreenPorch PoolArea  Fence  \
0            Lvl    AllPub  ...         0         120        0  MnPrv   
1            Lvl    AllPub  ...         0           0        0    NaN   
2            Lvl    AllPub  ...         0           0        0  MnPrv   
3            Lvl    AllPub  ...         0           0        0    NaN   
4            HLS    AllPub  ...         0         144        0    NaN   
...          ...       ...  ...       ...         ...      ...    ...   
1454         Lvl    AllPub  ...         0           0        0    NaN   
1455         Lvl    AllPub  ...         0           0        0    NaN   
1456         Lvl    AllPub  ...         0           0        0    NaN   
1457         Lvl    AllPub  ...         0           0        0  MnPrv   
1458         Lvl    AllPub  ...         0           0        0    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 79 columns]

here we will deal with alley (1369 missing points)

In [7]:
import numpy as np
data['Alley'] = data['Alley'].fillna('nan')
test_data['Alley'] = data['Alley'].fillna('nan')
np.unique(data["Alley"].values)

array(['Grvl', 'Pave', 'nan'], dtype=object)

In [8]:
Alley_mapping = {"nan": 0, "Grvl": 1, "Pave": 2}
data["Alley"] = data["Alley"].map(Alley_mapping)
test_data["Alley"] = test_data["Alley"].map(Alley_mapping)
np.unique(test_data["Alley"])

array([0, 1, 2], dtype=int64)

dealing with LotFrontage : i decided to assign 0 to the missing points (~250 missing points)

In [9]:
data['LotFrontage'] = data['LotFrontage'].fillna(0)
test_data["LotFrontage"] = test_data['LotFrontage'].fillna(0)

dealing with MasVnrArea and MasVnrType we have 8, 872 missed points respectively

In [10]:
data[data["MasVnrArea"].isna()]["MasVnrType"]

234     NaN
529     NaN
650     NaN
936     NaN
973     NaN
977     NaN
1243    NaN
1278    NaN
Name: MasVnrType, dtype: object

In [11]:
# the missing points in MasVnrArea are None in MasVnrType so we will substitute them with 0
data['MasVnrArea'] = data['MasVnrArea'].fillna(0)
test_data["MasVnrArea"] = test_data['MasVnrArea'].fillna(0)

In [12]:
# for MasVnrType -> one-hot encoding
data['MasVnrType'] = data['MasVnrType'].fillna('nan')
test_data['MasVnrType'] = data['MasVnrType'].fillna('nan')

In [13]:
df_encoded = pd.get_dummies(data, columns=['MasVnrType'], drop_first=True)
sale = df_encoded['SalePrice']
df_encoded = df_encoded.drop('SalePrice', axis=1)
df_encoded.insert(len(df_encoded.columns), "SalePrice", sale)
data = df_encoded
test_data = pd.get_dummies(test_data, columns=['MasVnrType'], drop_first=True)
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave      0      Reg   
1     1462          20       RL         81.0    14267   Pave      0      IR1   
2     1463          60       RL         74.0    13830   Pave      0      IR1   
3     1464          60       RL         78.0     9978   Pave      0      IR1   
4     1465         120       RL         43.0     5005   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1454  2915         160       RM         21.0     1936   Pave      2      Reg   
1455  2916         160       RM         21.0     1894   Pave      0      Reg   
1456  2917          20       RL        160.0    20000   Pave      0      Reg   
1457  2918          85       RL         62.0    10441   Pave      0      Reg   
1458  2919          60       RL         74.0     9627   Pave      0      Reg   

     LandContour Utilities  ...  Fence MiscFeature MiscVal MoSold YrSold  \
0            Lvl    AllPub  ...  MnPrv         NaN       0      6   2010   
1            Lvl    AllPub  ...    NaN        Gar2   12500      6   2010   
2            Lvl    AllPub  ...  MnPrv         NaN       0      3   2010   
3            Lvl    AllPub  ...    NaN         NaN       0      6   2010   
4            HLS    AllPub  ...    NaN         NaN       0      1   2010   
...          ...       ...  ...    ...         ...     ...    ...    ...   
1454         Lvl    AllPub  ...    NaN         NaN       0      6   2006   
1455         Lvl    AllPub  ...    NaN         NaN       0      4   2006   
1456         Lvl    AllPub  ...    NaN         NaN       0      9   2006   
1457         Lvl    AllPub  ...  MnPrv        Shed     700      7   2006   
1458         Lvl    AllPub  ...    NaN         NaN       0     11   2006   

     SaleType SaleCondition  MasVnrType_BrkFace  MasVnrType_Stone  \
0          WD        Normal                True             False   
1          WD        Normal               False             False   
2          WD        Normal                True             False   
3          WD        Normal               False             False   
4          WD        Normal                True             False   
...       ...           ...                 ...               ...   
1454       WD        Normal               False             False   
1455       WD       Abnorml               False             False   
1456       WD       Abnorml               False              True   
1457       WD        Normal               False             False   
1458       WD        Normal               False             False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 81 columns]

Deal with Bsmt_s
I did noticed that the same nombre of nan in ~ the same in all of them so i will see what its there?
so NA mean - No Basement

In [14]:
data.dtypes

Id                      int64
MSSubClass              int64
MSZoning               object
LotFrontage           float64
LotArea                 int64
                       ...   
SaleCondition          object
MasVnrType_BrkFace       bool
MasVnrType_Stone         bool
MasVnrType_nan           bool
SalePrice               int64
Length: 82, dtype: object

In [15]:
for col in data.columns:
    if not col.startswith("Bsmt"):
        continue
    if data[col].dtype == "object":
        data[col] = data[col].fillna("nan")
        test_data[col] = test_data[col].fillna("nan")
    else:
        data[col] = data[col].fillna(0)
        test_data[col] = test_data[col].fillna(0)

In [16]:
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0        1          60       RL         65.0     8450   Pave      0      Reg   
1        2          20       RL         80.0     9600   Pave      0      Reg   
2        3          60       RL         68.0    11250   Pave      0      IR1   
3        4          70       RL         60.0     9550   Pave      0      IR1   
4        5          60       RL         84.0    14260   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      0      Reg   
1456  1457          20       RL         85.0    13175   Pave      0      Reg   
1457  1458          70       RL         66.0     9042   Pave      0      Reg   
1458  1459          20       RL         68.0     9717   Pave      0      Reg   
1459  1460          20       RL         75.0     9937   Pave      0      Reg   

     LandContour Utilities  ... MiscFeature MiscVal MoSold YrSold SaleType  \
0            Lvl    AllPub  ...         NaN       0      2   2008       WD   
1            Lvl    AllPub  ...         NaN       0      5   2007       WD   
2            Lvl    AllPub  ...         NaN       0      9   2008       WD   
3            Lvl    AllPub  ...         NaN       0      2   2006       WD   
4            Lvl    AllPub  ...         NaN       0     12   2008       WD   
...          ...       ...  ...         ...     ...    ...    ...      ...   
1455         Lvl    AllPub  ...         NaN       0      8   2007       WD   
1456         Lvl    AllPub  ...         NaN       0      2   2010       WD   
1457         Lvl    AllPub  ...        Shed    2500      5   2010       WD   
1458         Lvl    AllPub  ...         NaN       0      4   2010       WD   
1459         Lvl    AllPub  ...         NaN       0      6   2008       WD   

     SaleCondition MasVnrType_BrkFace  MasVnrType_Stone  MasVnrType_nan  \
0           Normal               True             False           False   
1           Normal              False             False            True   
2           Normal               True             False           False   
3          Abnorml              False             False            True   
4           Normal               True             False           False   
...            ...                ...               ...             ...   
1455        Normal              False             False            True   
1456        Normal              False              True           False   
1457        Normal              False             False            True   
1458        Normal              False             False            True   
1459        Normal              False             False            True   

      SalePrice  
0        208500  
1        181500  
2        223500  
3        140000  
4        250000  
...         ...  
1455     175000  
1456     210000  
1457     266500  
1458     142125  
1459     147500  

[1460 rows x 82 columns]

In [17]:
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave      0      Reg   
1     1462          20       RL         81.0    14267   Pave      0      IR1   
2     1463          60       RL         74.0    13830   Pave      0      IR1   
3     1464          60       RL         78.0     9978   Pave      0      IR1   
4     1465         120       RL         43.0     5005   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1454  2915         160       RM         21.0     1936   Pave      2      Reg   
1455  2916         160       RM         21.0     1894   Pave      0      Reg   
1456  2917          20       RL        160.0    20000   Pave      0      Reg   
1457  2918          85       RL         62.0    10441   Pave      0      Reg   
1458  2919          60       RL         74.0     9627   Pave      0      Reg   

     LandContour Utilities  ...  Fence MiscFeature MiscVal MoSold YrSold  \
0            Lvl    AllPub  ...  MnPrv         NaN       0      6   2010   
1            Lvl    AllPub  ...    NaN        Gar2   12500      6   2010   
2            Lvl    AllPub  ...  MnPrv         NaN       0      3   2010   
3            Lvl    AllPub  ...    NaN         NaN       0      6   2010   
4            HLS    AllPub  ...    NaN         NaN       0      1   2010   
...          ...       ...  ...    ...         ...     ...    ...    ...   
1454         Lvl    AllPub  ...    NaN         NaN       0      6   2006   
1455         Lvl    AllPub  ...    NaN         NaN       0      4   2006   
1456         Lvl    AllPub  ...    NaN         NaN       0      9   2006   
1457         Lvl    AllPub  ...  MnPrv        Shed     700      7   2006   
1458         Lvl    AllPub  ...    NaN         NaN       0     11   2006   

     SaleType SaleCondition  MasVnrType_BrkFace  MasVnrType_Stone  \
0          WD        Normal                True             False   
1          WD        Normal               False             False   
2          WD        Normal                True             False   
3          WD        Normal               False             False   
4          WD        Normal                True             False   
...       ...           ...                 ...               ...   
1454       WD        Normal               False             False   
1455       WD       Abnorml               False             False   
1456       WD       Abnorml               False              True   
1457       WD        Normal               False             False   
1458       WD        Normal               False             False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 81 columns]

In [18]:
# here i will just get the target Column just to work comfortably.
sale = data['SalePrice']
data = data.drop('SalePrice', axis=1)

For the decoding we will use the mapping be cuz the order if meaningful

In [19]:
mapping = {
    "BsmtQual": {"Ex": 1, "Gd": 2, "TA": 3, "Fa": 4, "Po": 5, "nan": 6},
    "BsmtCond": {"Ex": 1, "Gd": 2, "TA": 3, "Fa": 4, "Po": 5, "nan": 6},
    "BsmtExposure": {"Gd": 1, "Av": 2, "Mn": 3, "No": 4, "nan": 5},
    "BsmtFinType1": {"GLQ": 1, "ALQ": 2, "BLQ": 3, "Rec": 4, "LwQ": 5, "Unf": 6, "nan": 7},
    "BsmtFinType2": {"GLQ": 1, "ALQ": 2, "BLQ": 3, "Rec": 4, "LwQ": 5, "Unf": 6, "nan": 7},
}
for col in mapping:
    data[col] = data[col].map(mapping[col])
    test_data[col] = test_data[col].map(mapping[col])